trace/doc.go

```diff
@@ -0,0 +1,160 @@
+// Package trace provies an API for initiating and reporting SSF
+// traces and attaching spans to them. Veneur's tracing API also
+// provides an experimental opentracing compatibility layer.
+//
+// The Veneur tracing API is completely independent of the opentracing
+// compatibility layer, with the exception of one convenience
+// function.
+//
+// Setup and Structure of the API
+//
+// To use the API, user code must set its service name via the global
+// variable Service, which is used by trace visualization tools to
+// tell apart the spans reported by different services. Code using the
+// trace API must set its Service, and set it only once, in main():
+//
+//    trace.Service = "my-awesome-service"
+//
+// The trace API is built on three abstractions: the
+// Client, the ClientBackend and the Span.
+//
+// Context
+//
+// Since it is very common for networked code to use context.Context,
+// veneur's trace API is designed to be used together with
+// Contexts. The API uses Contexts to store the current Span, and will
+// automatically pick the appropriate parent span from the Context.
+//
+// To effectively use the trace API, user code needs to ensure that
+// all functions that communicate externally (which includes using
+// traces) take Contexts as parameters, and pass them down.
+//
+// See https://blog.golang.org/context for an introduction to using
+// Context.
+//
+// Spans
+//
+// Spans represent the duration of a "unit of work" in a system: They
+// have a start and an end time, a name, and additional data that can
+// help identify what the work was. Example units of work could be the
+// handling of an HTTP request, or a database read. For convenience,
+// it is usually easiest to have a span represent a single function to
+// delineate the start and end, which makes it easy to create the Span
+// and report it via a defer statement.
+//
+// To create a new Span, use StartSpanFromContext. This function will
+// create a trace parented to the traces that came before it in the
+// Context. Typically, users should pass "" as the name to let
+// StartSpanFromContext figure out the correct name from caller
+// information.
+//
+//   span, ctx := trace.StartSpanFromContext(ctx, "")
+//   // report the span with the default client when this function returns:
+//   defer span.Finish()
+//   ... // do work here
+//
+// Reporting Spans via the trace Client
+//
+// Once constructed, SSF Spans must be reported to a system that
+// processes them. To do that, package trace exports a trace
+// Client. Typical applications will want to use the DefaultClient
+// exposed by this package. By default, it is set up to send spans to
+// veneur listening on the default SSF UDP port, 8128. Client code can
+// use SetDefaultClient to change the default client.
```

Comment:
Hm, I'm not sure it would fit very well there... I've rephrased this para a bit, though.

Reply:
Okay. To clarify I'm suggest consolidating the information related to why and how to set a default client. Having 2 separate bits of related information seems like it could be better organized together.

Reply:
OK, I think I want to say two things:

* Most applications should set a default client and use that.
* If someone wants to test what traces their app is sending, they need to keep handing around a concrete Client that can be substituted by test code; if so, they'll want to make use of a non-networked backend like the one we expose in #582.

This and #582 kinda depend on each other, so I think we'll need to polish this once more once both have landed.